### PR TITLE
HistogramData rollout: algorithms LoadMcStasNexus to LoadPreNexusMonitors

### LINKED ISSUE #18494: HistogramData Rollout: LoadMcStasNexus to LoadPreNexusMonitors
HistogramData rollout: algorithms <LoadMcStasNexus to LoadPreNexusMonitors>

This issue is part of the HistogramData rollout, see https://github.com/mantidproject/mantid/issues/17641.
Algorithms covered in this issue are:

The following algorithms were refactored to use HistogramData:
Framework/DataHandling/src/LoadMcStasNexus.cpp
Framework/DataHandling/src/LoadMLZ.cpp
Framework/DataHandling/src/LoadMuonNexus1.cpp
Framework/DataHandling/src/LoadMuonNexus2.cpp
Framework/DataHandling/src/LoadNexusMonitors2.cpp
Framework/DataHandling/src/LoadNexusProcessed.cpp
Framework/DataHandling/src/LoadNXcanSAS.cpp
Framework/DataHandling/src/LoadNXSPE.cpp
Framework/DataHandling/src/LoadPDFgetNFile.cpp
Framework/DataHandling/src/LoadPreNexusMonitors.cpp

### PR DESCRIPTION
**Pull request by LamarMoore**
## Description of work.
This PR is part of refactoring effort to use the new interface of the HistogramData module.
See https://github.com/mantidproject/mantid/issues/17641 for details.

The following algorithms were refactored to use HistogramData:

Performance tests were added for:

The following algorithms have improved performance:

## To test.

Code review.

Fixes #18494 .
Internal change, no release notes (potential performance improvements will be added to release notes as part of the umbrella issue).

### CODE REVIEW

File: Framework/DataHandling/test/LoadMLZTest.h [hunk 1]
```diff
@@ -69,15 +69,18 @@ class LoadMLZTestPerformance : public CxxTest::TestSuite {
 public:
   LoadMLZTestPerformance() : m_dataFile("TOFTOFTestdata.nxs") {}
```
Comment (SimonHeybrock) [inline Framework/DataHandling/test/LoadMLZTest.h]: Could you add create/destroy suite methods?

File: Framework/DataHandling/test/LoadNexusMonitorsTest.h [hunk 2]
```diff
@@ -240,4 +240,53 @@ class LoadNexusMonitorsTest : public CxxTest::TestSuite {
   }
 };
 
+class LoadNexusMonitorsTestPerformance : public CxxTest::TestSuite {
+public:
```
Comment (SimonHeybrock) [inline Framework/DataHandling/test/LoadNexusMonitorsTest.h]: suite methods?

**Review by SimonHeybrock — COMMENTED** [review]
A few comments below:

File: Framework/DataHandling/src/LoadNXcanSAS.cpp [hunk 3]
```diff
@@ -283,10 +278,14 @@ void read2DWorkspace(H5::DataSet &dataSet,
   hsize_t memSpaceDimension[1] = {dimInfo.dimBin};
   H5::DataSpace memSpace(1, memSpaceDimension);
 
+  auto &dataHist = func(workspace, 0);
+  Mantid::MantidVec data(dataHist.size());
   for (size_t index = 0; index < dimInfo.dimSpectrumAxis; ++index) {
     // Set the dataSpace to a 1D HyperSlab
     fileSpace.selectHyperslab(H5S_SELECT_SET, sizeOfSingleSlab, start);
-    dataSet.read(func(workspace, index), memoryDataType, memSpace, fileSpace);
```
Comment (LamarMoore) [inline Framework/DataHandling/src/LoadNXcanSAS.cpp]: No, `func(workspace, index)` was returning a pointer to a double array. Now the double array is passed directly. `func(workspace, index)` now returns a `HistogramData` type, so we just copy the data.

File: Framework/DataHandling/src/LoadNexusMonitors2.cpp [hunk 4]
```diff
@@ -491,53 +500,37 @@ void LoadNexusMonitors2::splitMutiPeriodHistrogramData(
   }
 
   // Y array should be divisible by the number of periods
-  if (m_workspace->blocksize() % numPeriods != 0) {
+  if (m_multiPeriodCounts[0].size() % numPeriods != 0) {
```
Comment (LamarMoore) [inline Framework/DataHandling/src/LoadNexusMonitors2.cpp]: See below explanation.

File: Framework/DataHandling/src/LoadNexusMonitors2.cpp [hunk 5]
```diff
@@ -491,53 +500,37 @@ void LoadNexusMonitors2::splitMutiPeriodHistrogramData(
...
 
-    // assign y values - use the values offset by the period number
     for (size_t wsIndex = 0; wsIndex < numSpectra; wsIndex++) {
-      MantidVec &outputVec = wsPeriod->dataY(wsIndex);
-      const MantidVec &inputVec = m_workspace->readY(wsIndex);
-      for (size_t index = 0; index < yLength; index++) {
-        outputVec[index] = inputVec[(yLength * i) + index];
-      }
-    }
+      auto inYBegin = m_multiPeriodCounts[wsIndex].cbegin() + offset;
```
Comment (LamarMoore) [inline Framework/DataHandling/src/LoadNexusMonitors2.cpp]: Before, `m_workspace` was used to store the multiperiod data which consisted of a number of histograms where the size of the y array would be at least twice the size of the x array. This violated the histogramdata type, so I stored the multiperiod data in `m_multiPeriodCounts/BinEdges`. The sizes of these `m_multiPeriodCounts` vectors correspond to the old size of `m_workspace->blocksize()`. The the calculation `m_multiPeriodCounts[0].size() % numPeriods` is identical. The code below is splitting each of the multiperiod histograms and storing them in the `wsPeriod` workspace which is designed to correctly store the multiperiod data. In essence, the workspace type was bastardized.

Comment (SimonHeybrock) [inline Framework/DataHandling/src/LoadNexusMonitors2.cpp]: @LamarMoore Maybe I misunderstand, but didn't you previously have, e.g., Y of length `nBin * nPeriod` and now you have `m_multiPeriodCounts` of length `nPeriod`, with each entry of length `nBin`?

Comment (LamarMoore) [inline Framework/DataHandling/src/LoadNexusMonitors2.cpp]: No, the `m_multiPeriodCounts` are of length `nPeriod * nBins` also. So I am just using this to replace what was being done with the workspace.

Comment (SimonHeybrock) [inline Framework/DataHandling/src/LoadNexusMonitors2.cpp]: Ah ok, I though you had split that compound index. IF not then this is ok :+1:

**Comment by SimonHeybrock** [comment]
I have no more comments, :shipit: once builds pass.

**Pull request merged**